File: matlab_files/python_substituted/associate_stress_coord.m
function [complete_files_path,spherical_complete_files_path] = ...
    associate_stress_coord(individual_element_paths,individual_path,...
    large_node_matrix_path,headers_on,stress_matrices_paths,coord_file)
% This function uses the results of both nodes_processor and
% elements_processor to calculate the centroid coordinates for each element
% and associate them to the corresponding element and stress components. 

% As before, define the relevant file paths and folders, creating them if
% they do not exist

individual_stress_paths = stress_matrices_paths;
dir_csv_elem_files = dir([individual_element_paths '\*.csv']);
csv_elem_files = {dir_csv_elem_files.name};
dir_csv_stress_files = dir([individual_stress_paths '\*.csv']);
csv_stress_files = {dir_csv_stress_files.name};
centroid_files_path = [individual_path '\Centroids'];
if ~exist(centroid_files_path, 'dir')
    mkdir (centroid_files_path)
end
complete_files_path = [individual_path '\Complete_Files'];
if ~exist(complete_files_path, 'dir')
    mkdir (complete_files_path)
end
spherical_centroids_files_path = [individual_path '\Spherical_Centroids'];
if ~exist(spherical_centroids_files_path, 'dir')
    mkdir (spherical_centroids_files_path)
end
spherical_components_files_path = [individual_path '\Spherical_Components'];
if ~exist(spherical_components_files_path, 'dir')
    mkdir (spherical_components_files_path)
end
spherical_complete_files_path = [individual_path '\Spherical_Complete_Files'];
if ~exist(spherical_complete_files_path, 'dir')
    mkdir (spherical_complete_files_path)
end

% Define headers if we want them in the main function

complete_headers = {'Label', 'S_Mises', 'S_11', 'S_22', 'S_33', 'S_12', 'S_13', 'S_23', ...
     'X_centroid', 'Y_centroid', 'Z_centroid'};
centroid_headers = {'Centr_label', 'X_centroid', 'Y_centroid', 'Z_centroid'};

% Initialize necessary variables

all_nodes = readmatrix(large_node_matrix_path);
large_centroids = [];
large_complete_file = [];
spherical_earth_point_large_complete = [];


% The first external for cycle runs over the list of csv files contaning
% the stress components for each part and region. However, some adjustments
% have to be made. Because the region stress files all refer to a single
% larger part element file in EARTH_POINT and have similar names, every
% time a file is processed its name is removed from the list of stress
% files names. Then, the first file in the list is always processed, since
% the previous ones have been deleted from the list of possible files to
% read.

for i = 1:length(csv_stress_files)
    size_to_evaluate = dir([dir_csv_stress_files(i).folder '\' csv_stress_files{i}]);
    if size_to_evaluate.bytes < 100
        csv_stress_files{i} = {};
    end
end
csv_stress_files = csv_stress_files(~cellfun(@isempty,csv_stress_files));

% Verify if the files are already all there based on whether the last file
% to be created is actually already in the folder

if isfile([spherical_complete_files_path '\Spherical_complete_file_' coord_file(1:5) '.csv'])
    disp(['The files containing centroid stresses associated to the relative coordinates' ...
        ' already exist, moving on to classification of stress values based on depth.'])
else 
    disp('Associating stress components to corresponding centroids...')
    for i = 1:length(csv_stress_files)
        % Here we create the list of stress files and extract the first file in
        % the list, which is the one to read
        
        csv_stress_files = csv_stress_files(~cellfun('isempty',csv_stress_files));
        csv_elem_files = csv_elem_files(~cellfun('isempty',csv_elem_files));
        part_file = csv_stress_files{1};
        disp(['Processing the following stress file: ', part_file]);
        part_matrix = readmatrix([individual_stress_paths '\' part_file]);
        
        % Here we find the name of the stress file to open and read
        
        region_files_number = nnz(find(contains(csv_stress_files,'Region')));
        region_flag = contains(part_file,'Region');
        if region_flag == 1
            file_to_read_logical = contains(extractAfter(csv_elem_files,'Part_'),extractBefore(part_file,'_Region')) == 1;
        else
            file_to_read_logical = strcmp(extractAfter(csv_elem_files,'Part_'),part_file) == 1;
        end
        file_to_read_index = (file_to_read_logical == 1);
        
        % Read all elements for a certain part and allocate a matrix where
        % to store centroid coordinates
        all_elems = readmatrix([individual_element_paths '\' csv_elem_files{file_to_read_index}]);
        centroid_coord = zeros(size(part_matrix,1),4);
        % complete_file = zeros(size(part_matrix,1),4);
        % This for cycle is used to calculate the centroid coordinates for each
        % element of each part. The external for cycle loops over the elements
        % of that part.
        
        for j=1:size(part_matrix,1)
            % Here we use the element label from the stress matrix of the part
            % or region to find its node labels based on its label in the
            % element matrix of the part, then we read the nodes for that
            % element
            
            elem_nodes_index = all_elems(:,1) == part_matrix(j,1);
            elem_nodes = all_elems(elem_nodes_index,2:end);
            
            % 6-node elements will have two zeros at the end that we will take
            % out, and we can do this because we process on row at a time
            
            elem_nodes = nonzeros(elem_nodes);
            
            % Allocate a matrix where to store the three coordinates of each node
            % of the considered element, which means an 8- or 6-by-3 matrix
            % depending on the element type itself, then calculate the
            % centroid coordinates as the mean of the coordinates of each
            % node
            
            elem_coord_matrix = zeros(length(elem_nodes),3);
            
            for k = 1:size(elem_nodes,1)
                index = all_nodes(:,1) == elem_nodes(k,1);
                nodes_coord = all_nodes(index,2:4);
                if ~isempty(nodes_coord)
                    elem_coord_matrix(k,1:end) = nodes_coord;
                end
            end
            for l = 1:3
                centroid_coord(j,l+1) = mean(elem_coord_matrix(1:end,l));
            end
            centroid_coord(j,1) = part_matrix(j,1);
            if centroid_coord(j,2:4) == 0
                centroid_coord(j,:) = [];
            end
        end
        
        % Swap the Y and Z coordinates because ABAQUS uses another
        % coordinate system, and for the same reason swap the stress
        % components around 
        
        centroid_coord(:, [3 4]) = centroid_coord(:, [4 3]);
        centroid_coord(:, [2 3]) = centroid_coord(:, [3 2]);
        part_matrix(:, [3 5]) = part_matrix(:, [5 3]);
        part_matrix(:, [3 4]) = part_matrix(:, [4 3]);
        part_matrix(:, [7 6]) = part_matrix(:, [6 7]);
        part_matrix(:, [6 8]) = part_matrix(:, [8 6]);
        
        % Write both the centroid coordinates and the complete file with
        % centroid coordinates and stress components for each region, then
        % delete the name of the file that has just been read from the list of
        % stress csv files.
        
        complete_file = [part_matrix,centroid_coord(:,2:4)];
        if (region_flag == 1)
            large_centroids = [large_centroids; centroid_coord];
            large_complete_file = [large_complete_file; complete_file];
        end
        
        centroid_individual_path = [centroid_files_path '\Centroid_file_' part_file];
        complete_individual_path = [complete_files_path '\Complete_file_' part_file];
        
        % As before, determine how to save the matrix based on the presence
        % of headers or not
        
        if headers_on == 1
            complete_file_headers = [complete_headers; num2cell(complete_file)];
            centroid_coord_headers = [centroid_headers; num2cell(centroid_coord)];
            writecell(complete_file_headers,complete_individual_path);
            writecell(centroid_coord_headers,centroid_individual_path);
        else
            writematrix(complete_file,complete_individual_path);
            writematrix(centroid_coord,centroid_individual_path);
        end
        
        % Here the rotation of the stress tensor into geographical coordinates
        % is performed.
        complete_file_spherical = rotate_tensor(part_matrix,spherical_centroids_files_path, ...
            spherical_components_files_path,spherical_complete_files_path,headers_on,...
            complete_headers,centroid_headers,complete_individual_path,part_file);
        
        % Create the complete matrix containing the spherical stress components
        % for part Earth too, needed for the depth classification.
        if (region_flag == 1)
            spherical_earth_point_large_complete = [spherical_earth_point_large_complete; complete_file_spherical];
        end
        
        csv_stress_files{1} = [];
        if region_flag == 0
            csv_elem_files{file_to_read_logical} = [];
        elseif region_files_number == 0
            csv_elem_files{file_to_read_logical} = [];
        end
    end
    
    if headers_on == 1
        large_centroids = [centroid_headers; num2cell(large_centroids)];
        large_complete_file = [complete_headers; num2cell(large_complete_file)];
        spherical_earth_point_large_complete = [complete_headers; num2cell(spherical_earth_point_large_complete)];
        writecell(large_centroids,[centroid_files_path '\Centroid_coordinates_' coord_file(1:5) '.csv']);
        writecell(large_complete_file,[complete_files_path '\Complete_file_' coord_file(1:5) '.csv']);
        writecell(spherical_earth_point_large_complete,[spherical_complete_files_path '\Spherical_complete_file_' coord_file(1:5) '.csv']);
    else
        writematrix(large_centroids,[centroid_files_path '\Centroid_coordinates_' coord_file(1:5) '.csv']);
        writematrix(large_complete_file,[complete_files_path '\Complete_file_' coord_file(1:5) '.csv']);
        writematrix(spherical_earth_point_large_complete,[spherical_complete_files_path '\Spherical_complete_file_' coord_file(1:5) '.csv']);
    end
    
end

end

File: python_files_python_3/associate_stress_coord.py
from region_counter import *
from rotate_tensor import *
from cart2geo import *
import pandas as pd


def associate_stress_coord(individual_element_paths, stress_part_values, large_node_matrix_path, headers_on,
                           stress_matrices_path, coord_file, coupled_stress_folder):
    # This function associates each stress value vector to the corresponding centroid coordinates, both of them in
    # Cartesian and geographical reference systems, then saves them in separate .csv files for each part.
    import os
    import numpy as np
    import csv
    import pdb

    # Define the extension we have to read data from and gather the csv names containing the element nodes and stress
    # values to associate
    file_extension = '.csv'
    individual_stress_paths = stress_matrices_path
    dir_csv_elem_files = os.listdir(individual_element_paths)
    dir_csv_stress_files = os.listdir(individual_stress_paths)
    csv_elem_files = [filename for filename in dir_csv_elem_files if filename.endswith(file_extension)]
    csv_stress_files = [filename for filename in dir_csv_stress_files if filename.endswith(file_extension)]

    # Define the relevant folders to save centroid coordinates and complete tables and create them if they do not
    # exist yet
    centroid_files_path = os.path.join(stress_part_values, 'Centroids')
    if not os.path.exists(centroid_files_path):
        os.mkdir(centroid_files_path)
    complete_files_path = os.path.join(stress_part_values, 'Complete_files')
    if not os.path.exists(complete_files_path):
        os.mkdir(complete_files_path)
    geographical_complete_files_path = os.path.join(stress_part_values, 'Geographical_complete_files')
    if not os.path.exists(geographical_complete_files_path):
        os.mkdir(geographical_complete_files_path)

    # Define headers for the complete files and the centroid files
    complete_headers = ['Label', 'S_Mises', 'S_11', 'S_22', 'S_33', 'S_12', 'S_13', 'S_23', 'X_centroid', 'Y_centroid',
                        'Z_centroid', 'R', 'Depth', 'Lat', 'Lon']
    centroid_headers = ['Centr_label', 'X_centroid', 'Y_centroid', 'Z_centroid']

    # Initialize lists to fill
    large_centroids = []
    large_complete_file = []

    # For all the next operations we will have to refer to the large matrix containing all the part nodes, because
    # for each part and for each element we will have to extract the coordinates of each node

    # Decide whether to run the main algorithm or not based on the presence of the last file to be generated
    if os.path.isfile(os.path.join(complete_files_path, 'Stress_association_completion_certificate.txt')):
        print('The files containing centroid stresses associated to the relative coordinates already exist, moving on '
              'to classification of stress values based on depth.')
    else:
        print('Associating stress components to corresponding centroids...')
        with open(large_node_matrix_path) as file_to_read:
            # Read the matrix containing the list of nodes for all parts
            all_nodes = pd.read_csv(file_to_read, delimiter=",")

            # Filter out unwanted parts and if some matrices are completely empty, do not process them
            for file_to_evaluate in range(len(csv_stress_files)):
                if 'Region' in csv_stress_files[file_to_evaluate]:
                    csv_stress_files[file_to_evaluate] = 'to_delete'
                if 'LOW' in csv_stress_files[file_to_evaluate]:
                    csv_stress_files[file_to_evaluate] = 'to_delete'
                if 'I0' in csv_stress_files[file_to_evaluate]:
                    csv_stress_files[file_to_evaluate] = 'to_delete'
                if 'I1' in csv_stress_files[file_to_evaluate]:
                    csv_stress_files[file_to_evaluate] = 'to_delete'
            csv_stress_files = [a for a in csv_stress_files if a != 'to_delete']
            for file_to_evaluate in range(len(csv_stress_files)):
                with open(os.path.join(individual_stress_paths, csv_stress_files[file_to_evaluate])) as matrix:
                    matrix_to_evaluate = pd.read_csv(matrix, delimiter=",")
                    if len(matrix_to_evaluate) == 0:
                        csv_stress_files[file_to_evaluate] = 'to_delete'

        csv_stress_files = [a for a in csv_stress_files if a != 'to_delete']
        # Iterate over the files containing the stress values for each part or region
        for i in range(len(csv_stress_files)):
            # Eliminate the empty elements from the list at each iteration, because after we process each file they
            # will be taken away from the list
            csv_stress_files = [_f for _f in csv_stress_files if _f]
            csv_elem_files = [_f for _f in csv_elem_files if _f]
            # Select the stress file to process
            part_file = csv_stress_files[0]
            print('Processing the following stress file: ' + part_file)

            file_to_read_logical = [i.split('Elements_Part_')[1] == part_file for i in csv_elem_files]
            file_to_read_index = np.array([file_to_read_logical.index(i) for i in file_to_read_logical
                                           if i == True])
            # Open the element file and transform both the element and stress files into arrays for easier handling
            file_to_read_index = file_to_read_index[0]
            all_elems_to_open = open(os.path.join(individual_element_paths, csv_elem_files[file_to_read_index]))
            # all_elems = all_elems_to_open.readlines()
            # all_elems = [line.strip() for line in all_elems[1:]]
            # all_elems = [np.array([eval(i) for i in line.split(",")[:]]) for line in all_elems]
            # all_elems = np.array(all_elems)
            all_elems = pd.read_csv(all_elems_to_open, delimiter=",")
            part_matrix_to_open = open(os.path.join(individual_stress_paths, part_file))
            part_matrix = pd.read_csv(part_matrix_to_open, delimiter=",")
            centroid_coord = np.zeros((len(part_matrix), 4))

            # For each element, fill a vector containing all of its nodes; extract the coordinates for each of them,
            # then calculate the mean across these nodes for all three coordinates
            for j in range(len(part_matrix)):
                elem_nodes_index = [np.where(all_elems.iloc[:, 0] == part_matrix.iloc[j, 0])]
                elem_nodes = all_elems.iloc[elem_nodes_index, 1:]
                elem_nodes = elem_nodes.iloc[np.nonzero(elem_nodes)]
                elem_coord_matrix = np.zeros((len(elem_nodes), 3))
                for k in range(0, len(elem_nodes)):
                    index = [np.where(all_nodes.iloc[:, 0] == elem_nodes.iloc[k])]
                    nodes_coord = all_nodes.iloc[index, 1:]
                    if not len(nodes_coord) == 0:
                        elem_coord_matrix[k, :] = nodes_coord
                for m in range(0, 3):
                    centroid_coord[j, m+1] = np.mean(elem_coord_matrix[:, m])
                centroid_coord[j, 0] = part_matrix[j, 0]
                if np.all(centroid_coord[j, 1:]) == 0:
                    centroid_coord = np.delete(centroid_coord, j, axis=0)

            # Because of ABAQUS' reference system, swap the Y and Z coordinate first and then the X and Y
            # coordinates

            centroid_coord[:, 1], centroid_coord[:, 3] = centroid_coord[:, 3].copy(), centroid_coord[:, 1].copy()
            centroid_coord[:, 2], centroid_coord[:, 3] = centroid_coord[:, 3].copy(), centroid_coord[:, 2].copy()
            part_matrix[:, 2], part_matrix[:, 4] = part_matrix[:, 4].copy(), part_matrix[:, 2].copy()
            part_matrix[:, 3], part_matrix[:, 4] = part_matrix[:, 4].copy(), part_matrix[:, 3].copy()
            part_matrix[:, 5], part_matrix[:, 7] = part_matrix[:, 7].copy(), part_matrix[:, 5].copy()
            part_matrix[:, 5], part_matrix[:, 6] = part_matrix[:, 6].copy(), part_matrix[:, 5].copy()

            # Create the complete file for each part and region containing the stresses and centroid coordinates,
            # appending each region file of the Earth part to the Earth large matrix so that we can also have the
            # values for the part and not just for its sub-regions
            complete_file = np.hstack((part_matrix, centroid_coord[:, 1:]))
            earth_radius = 6371000
            depths = np.zeros((len(complete_file), 1))
            radial_centroid_distance = np.zeros((len(complete_file), 1))
            for k in range(len(complete_file)):
                radial_centroid_distance[k] = np.sqrt(complete_file[k, -3] ** 2 + complete_file[k, -2] ** 2 +
                                                      complete_file[k, -1] ** 2)
                depths[k] = earth_radius - radial_centroid_distance[k]
            complete_file = np.column_stack((complete_file, radial_centroid_distance, depths))
            cartesian_coordinates = complete_file[:, 8:11]
            # Calculated the geographical coordinates and append them to the previous matrix
            [lat, lon] = cart2geo(cartesian_coordinates)
            complete_file = np.column_stack((complete_file, lat, lon))

            # Define name used to save the files
            centroid_individual_path = os.path.join(centroid_files_path, 'Centroid_file_' + part_file)
            complete_individual_path = os.path.join(complete_files_path, 'Complete_file_' + part_file)
            # Write the centroid coordinates and the complete files in csvs with headers or not
            if headers_on == 1:
                with open(centroid_individual_path, 'wb') as f_write:
                    writer = csv.writer(f_write)
                    writer.writerow(centroid_headers)
                    writer.writerows(centroid_coord)
                with open(complete_individual_path, 'wb') as f_write:
                    writer = csv.writer(f_write)
                    writer.writerow(complete_headers)
                    writer.writerows(complete_file)
            else:
                with open(centroid_individual_path, 'wb') as f_write:
                    writer = csv.writer(f_write)
                    writer.writerows(centroid_coord)
                with open(complete_individual_path, 'wb') as f_write:
                    writer = csv.writer(f_write)
                    writer.writerows(complete_file)

            # Calculate and save the geographical stress components
            rotate_tensor(part_matrix, geographical_complete_files_path, headers_on, complete_headers,
                          complete_individual_path, part_file)

        # Only save
        for csv_file in range(len(csv_stress_files)):
            if csv_stress_files[csv_file] == 'EARTH.csv':
                if not os.path.exists(os.path.join(complete_files_path, 'Geographical_complete_file_coupled_' +
                                                                        csv_stress_files[csv_file])):
                    with open(os.path.join(complete_files_path, 'Complete_file_' + csv_stress_files[csv_file]),
                              'r') as uncoupled_coordinate_read_obj:
                        uncoupled_coordinate_matrix = uncoupled_coordinate_read_obj.readlines()
                        uncoupled_coordinate_matrix = [line.strip() for line in uncoupled_coordinate_matrix[1:]]
                        uncoupled_coordinate_matrix = [np.array([eval(i) for i in line.split(",")[:]])
                                                       for line in uncoupled_coordinate_matrix]
                        uncoupled_coordinate_matrix = np.array(uncoupled_coordinate_matrix)
                    with open(os.path.join(coupled_stress_folder, csv_stress_files[csv_file]), 'r') as coordinate_read_obj:
                        individual_coordinate_matrix = coordinate_read_obj.readlines()
                        individual_coordinate_matrix = [line.strip() for line in individual_coordinate_matrix[1:]]
                        individual_coordinate_matrix = [np.array([eval(i) for i in line.split(",")[:]]) for line in
                                                        individual_coordinate_matrix]
                        individual_coordinate_matrix = np.array(individual_coordinate_matrix)
                    final_coupled_matrix = np.hstack((individual_coordinate_matrix, uncoupled_coordinate_matrix[:, 8:]))
                    complete_individual_path_coupled = os.path.join(complete_files_path, 'Complete_file_coupled_' +
                                                                    csv_stress_files[csv_file])
                    with open(complete_individual_path_coupled, 'wb') as f_write:
                        writer = csv.writer(f_write)
                        writer.writerow(complete_headers)
                        writer.writerows(final_coupled_matrix)
                    rotate_tensor(final_coupled_matrix, geographical_complete_files_path, headers_on, complete_headers,
                                  complete_individual_path_coupled, 'coupled_' + csv_stress_files[csv_file])

        with open(os.path.join(complete_files_path, 'Stress_association_completion_certificate.txt'), 'wb') as f_write:
            f_write.write('Stress association completed.')

    return centroid_files_path, complete_files_path, geographical_complete_files_path


